vllm/spec_decode/multi_step_worker.py

```diff
@@ -153,6 +153,12 @@ def _expand_execute_model_request(
 
         updated_execute_model_req.seq_group_metadata_list =\
             updated_seq_group_metadata_list
+
+        if isinstance(updated_execute_model_req.previous_hidden_states,
+                      HiddenStates):
+            updated_execute_model_req.previous_hidden_states\
```

Comment:
In this method we are changing the order of sequences in updated_execute_model_req compared to the original execute_model_req depending on whether sequences have bonus tokens or not. Wondering if order of the hidden states need to be aligned with the new ordering in updated_execute_model_req and if it is being handled in expand_with_bonus_tokens? E.g. in expand_with_bonus_tokens I see we are adding the hidden states for the bonus tokens before the one without the bonus token which is different compared to the order of seq_group_metadata in updated_execute_model_req. Is that expected?

Reply:
Actually, in expand_with_bonus_tokens, the hidden states for tokens before the bonus are added first since bonus_token_previous_hidden_states contains the hidden states of the token before the bonus. Maybe I should rename the variable name to avoid confusion.

Reply:
You are right, in this particular case, the order will not match. But HiddenStates init has an assertion that forces all seq groups to have only one sequence. So, I wrote the code with that assumption in mind. However I can change this, it shouldn't be a big change.

Yes, I can add the test as well but it should be ok to test with this assertion in mind, right? Or should we force-test the case where seq_group can have multiple sequences?

Reply:
Thanks for adding the test. Is there any existing test which would validate the assertion that each seq_group has one sequence. If not I was wondering if you could add a small test in https://github.com/vllm-project/vllm/blob/main/tests/test_sequence.py to validate this. I don't think the current test that you added would fail if we were to remove that assertion.

Reply:
I thought this was a broader assumption as we don't have support for beam search in speculative decoding. In that case, do you mean we should add a test to ensure that in speculative decoding each seq_group has one sequence?

Reply:
Sorry for the confusion. No need for this test.